### PR TITLE
perf: Effectively remove the use of self.astensor() in tensor backend calls

### LINKED ISSUE #295: Reduce the use of self.astensor() in tensor backend calls
# Description

Many of the calls in the backends wrap the `tensor_in` in `self.astensor` for convenience. This will probably result in some sloppy code and unnecessary conversions in the future. This issue is filed to remind us to clean this portion up.

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #295

Remove or reduce the guarding `self.astensor()` calls in the methods of the tensor backends. This makes it more explicit that tensors are required, move the responsibility to the user to format their data correctly, and should add a speedup by dramatically reducing the number of `self.astensor()` calls.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* perf: Remove all non-essential 'self.astensor' calls from the tensor backend methods
* Replace custom instances of 'where' with tensor library versions for TensorFlow, PyTorch, and MXNet backends
   - Uses: tf.compat.v2.where, torch.where, nd.where
* Make masks used in 'tensorlib.where' explicit bools
* docs: Update the docstring examples for the tensor backends
* test: Make all relevant arguments tensors in test_tensor.py
```

### CODE REVIEW

File: src/pyhf/interpolators/code0.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def __call__(self, alphasets):
         tensorlib, _ = get_backend()
         self._precompute_alphasets(tensorlib.shape(alphasets))
         where_alphasets_positive = tensorlib.where(
-            alphasets > 0, self.mask_on, self.mask_off
+            tensorlib.astensor(alphasets > 0), self.mask_on, self.mask_off
```
Comment (lukasheinrich) [inline src/pyhf/interpolators/code0.py]: why are we *adding* `astensor` calls here?

Comment (matthewfeickert) [reply src/pyhf/interpolators/code0.py]: In short, because I removed the `mask = self.astensor(mask)` from the `where` method in the TensorFlow backend. As it requires that the `mask` argument is a tensor I decided to add `tensorlib.astensor` calls to a few places where the mask is a logical requirement and not already a tensor given that

```
$ git grep "tensorlib.where(" src/ | wc -l
26
```

is more then the number of calls that I added. So it seems that explicitly calling it as needed is preferable then having the backends call it everytime.

Comment (matthewfeickert) [reply src/pyhf/interpolators/code0.py]: Alternatively, I could say that as this is a TensorFlow problem which might improve in TensorFlow 2.0 I could remove the calls to `tensorlib.astensor` that I added and just have TensorFlow do the redundant calls, until it might be possible to turn it off entirely.

Comment (lukasheinrich) [reply src/pyhf/interpolators/code0.py]: could we skip this if we instead assured that alphasets is a tensor to start with?

Comment (matthewfeickert) [reply src/pyhf/interpolators/code0.py]: > could we skip this if we instead assured that alphasets is a tensor to start with?

No. The issue is that `alphasets > 0` is being passed in. TensorFlow 1.0 needs to be told that this is a bool tensor with a `astensor` call.

Comment (matthewfeickert) [reply src/pyhf/interpolators/code0.py]: The decision is if we're going to make a few calls across all backends (which we are currently already doing) in the interpolator codes and hypotest, or if we are going to isolate all of those calls to the TensorFlow backend and deal with them later in the TensorFlow 2.0 migration.

I think we can probably do the latter.

**Comment by matthewfeickert** [comment]
The only `selt.astensor` that I think we don't need but I haven't figured out how to get is in

https://github.com/diana-hep/pyhf/blob/86269a013121d82d33662753caa3dfad7ed6dcd6/src/pyhf/tensor/pytorch_backend.py#L97-L99

[`torch.gather`](https://pytorch.org/docs/stable/torch.html#torch.gather) is confusing to me though, so I'm not sure how to properly get this.

File: src/pyhf/tensor/common.py [hunk 2]
```diff
@@ -45,9 +45,15 @@ def stitch(self, data):
     def split(self, data):
         tensorlib, _ = get_backend()
         if len(tensorlib.shape(data)) == 1:
-            return [tensorlib.gather(data, idx) for idx in self.partition_indices]
+            return [
+                tensorlib.gather(data, tensorlib.astensor(idx, dtype='int'))
```
Comment (lukasheinrich) [inline src/pyhf/tensor/common.py]: this can probaly go into `_precompute` since converting the indices into tensors is not dependent on `data`

File: src/pyhf/tensor/common.py [hunk 3]
```diff
@@ -45,9 +45,15 @@ def stitch(self, data):
...
+                for idx in self.partition_indices
+            ]
         data = tensorlib.einsum('...j->j...', tensorlib.astensor(data))
         return [
-            tensorlib.einsum('j...->...j', tensorlib.gather(data, idx))
+            tensorlib.einsum(
+                'j...->...j',
+                tensorlib.gather(data, tensorlib.astensor(idx, dtype='int')),
+            )
```
Comment (lukasheinrich) [inline src/pyhf/tensor/common.py]: dito

File: src/pyhf/tensor/common.py (hunk 2 above)
Comment (matthewfeickert) [reply src/pyhf/tensor/common.py]: Okay, I'll look into moving this. A note, this is _only_ being done to remove the `self.astensor` call in the PyTorch backend `gather` method (https://github.com/diana-hep/pyhf/pull/582/commits/2a163e71dc448494415960633faf72a6af1c974a). So this is a bit of a tradeoff, as this now requires extra ops by all other backends but it reduces the number of times that the PyTorch backend will have to call `astensor` overall.

File: src/pyhf/tensor/tensorflow_backend.py [hunk 4]
```diff
@@ -195,26 +186,20 @@ def reshape(self, tensor, newshape):
...
 
     def stack(self, sequence, axis=0):
         return tf.stack(sequence, axis=axis)
 
     def where(self, mask, tensor_in_1, tensor_in_2):
-        mask = self.astensor(mask)
-        tensor_in_1 = self.astensor(tensor_in_1)
-        tensor_in_2 = self.astensor(tensor_in_2)
-        return mask * tensor_in_1 + (1 - mask) * tensor_in_2
+        mask = tf.cast(mask, tf.bool)
```
Comment (kratsg) [inline src/pyhf/tensor/tensorflow_backend.py]: why is there a cast here? The mask should be `bool` already.

File: src/pyhf/tensor/tensorflow_backend.py [hunk 5]
```diff
@@ -195,26 +186,20 @@ def reshape(self, tensor, newshape):
...
+        return tf.compat.v2.where(mask, tensor_in_1, tensor_in_2)
```
Comment (kratsg) [inline src/pyhf/tensor/tensorflow_backend.py]: Can we always use latest, or does `tf` only version this method?

Comment (matthewfeickert) [reply src/pyhf/tensor/tensorflow_backend.py]: > Can we always use latest, or does `tf` only version this method?

@kratsg Sorry, I'm not quite sure what you mean by "latest" here, but if I guess a bit the reason I'm using `tf.compat.v2.where` is because it is [the alias](https://www.tensorflow.org/api_docs/python/tf/where) that TensorFlow recommends when you try to use `tf.where` while using the TF v1.0 API.

File: src/pyhf/tensor/tensorflow_backend.py (hunk 4 above)
Comment (matthewfeickert) [reply src/pyhf/tensor/tensorflow_backend.py]: > why is there a cast here? The mask should be `bool` already.

I think that it is better to just cast it to make sure. There are times like the following where the mask seems like a valid `Bool`

https://github.com/diana-hep/pyhf/blob/c8a94f4d5a11edfdd5a28184153e1ef4b076e638/tests/test_tensor.py#L48-L50

but TensorFlow doesn't see it as such if the `tf.cast(mask, tf.bool)` is removed. So I think that it should just be used to ensure the right Tensor `dtype`.

Comment (kratsg) [reply src/pyhf/tensor/tensorflow_backend.py]: https://github.com/diana-hep/pyhf/blob/d9aeac8bd3b7bd384342ae2e0e7547b78714e66c/src/pyhf/tensor/tensorflow_backend.py#L124

That's because the default cast is to float. You should use `tb.astensor(tensor, dtype='bool')` for the mask input.

File: src/pyhf/tensor/pytorch_backend.py [hunk 6]
```diff
@@ -131,35 +123,25 @@ def zeros(self, shape):
...
 
     def stack(self, sequence, axis=0):
         return torch.stack(sequence, dim=axis)
 
     def where(self, mask, tensor_in_1, tensor_in_2):
-        mask = self.astensor(mask).type(torch.FloatTensor)
-        tensor_in_1 = self.astensor(tensor_in_1)
-        tensor_in_2 = self.astensor(tensor_in_2)
-        return mask * tensor_in_1 + (1 - mask) * tensor_in_2
+        return torch.where(mask.type(torch.BoolTensor), tensor_in_1, tensor_in_2)
```
Comment (matthewfeickert) [inline src/pyhf/tensor/pytorch_backend.py]: @kratsg Your comment got removed when I combined the two lines, so answering here again:

> why is there a cast here? The mask should be `bool` already.

I think that it is better to just cast it to make sure. There are times like the following where the mask seems like a valid `Bool`

https://github.com/diana-hep/pyhf/blob/c8a94f4d5a11edfdd5a28184153e1ef4b076e638/tests/test_tensor.py#L48-L50

but TensorFlow doesn't see it as such if the `tf.cast(mask, tf.bool)` is removed. So I think that it should just be used to ensure the right Tensor `dtype`.

File: src/pyhf/tensor/tensorflow_backend.py (hunk 4 above)
Comment (kratsg) [reply src/pyhf/tensor/tensorflow_backend.py]: tl;dr - the test is wrong.

File: src/pyhf/tensor/tensorflow_backend.py [hunk 7]
```diff
@@ -195,26 +186,19 @@ def reshape(self, tensor, newshape):
...
 
     def stack(self, sequence, axis=0):
         return tf.stack(sequence, axis=axis)
 
     def where(self, mask, tensor_in_1, tensor_in_2):
-        mask = self.astensor(mask)
-        tensor_in_1 = self.astensor(tensor_in_1)
-        tensor_in_2 = self.astensor(tensor_in_2)
-        return mask * tensor_in_1 + (1 - mask) * tensor_in_2
+        return tf.compat.v2.where(tf.cast(mask, tf.bool), tensor_in_1, tensor_in_2)
```
Comment (matthewfeickert) [inline src/pyhf/tensor/tensorflow_backend.py]: @kratsg Your comment got removed when I combined two lines, so answering here again.

> why is there a cast here? The mask should be `bool` already.

I think that it is better to just cast it to make sure. There are times like the following where the mask seems like a valid `Bool`

https://github.com/diana-hep/pyhf/blob/c8a94f4d5a11edfdd5a28184153e1ef4b076e638/tests/test_tensor.py#L48-L50

but TensorFlow doesn't see it as such if the `tf.cast(mask, tf.bool)` is removed. So I think that it should just be used to ensure the right Tensor `dtype`.

Comment (kratsg) [reply src/pyhf/tensor/tensorflow_backend.py]: https://github.com/diana-hep/pyhf/blob/d9aeac8bd3b7bd384342ae2e0e7547b78714e66c/src/pyhf/tensor/tensorflow_backend.py#L124

That's because the default cast is to float. You should use `tb.astensor(tensor, dtype='bool')` for the mask input.

tl;dr - the test is wrong.

Comment (kratsg) [reply src/pyhf/tensor/tensorflow_backend.py]:
```
>>> import pyhf
>>> pyhf.set_backend(pyhf.tensor.tensorflow_backend())
>>> from pyhf import tensorlib as tb
>>> tb.astensor([1,0,1])
array([1., 0., 1.])
>>> tb.astensor([1,0,1], dtype='float')
array([1., 0., 1.])
>>> tb.astensor([1,0,1], dtype='bool')
array([ True, False,  True])
```

File: src/pyhf/tensor/tensorflow_backend.py (hunk 4 above)
Comment (matthewfeickert) [reply src/pyhf/tensor/tensorflow_backend.py]: > That's because the default cast is to float

Yes, I wasn't confused as to the _why_ of the behavior, but as to the reasonableness of it. I will change the test, but if you were a user it would seem to me that `tb.astensor([1, 0, 1])` could be a valid way to indicate a Bool mask. So I would opt to include that cast there, which should be faster then the `self.astensor()` call that we were previously including.

File: src/pyhf/tensor/tensorflow_backend.py (hunk 7 above)
Comment (matthewfeickert) [reply src/pyhf/tensor/tensorflow_backend.py]: Not sure where my last comment went (somewhere into the void?) but yes, it is clear as to the _why_ of the behavior by not to the reasonableness. I have changed the test, but I think it is better to keep the cast there as a user could quite legitimately view the `tb.astensor([1,0,1])` as a valid form of Bool (as NumPy does). The cast should be faster then the `self.astensor` call that we have previously been doing, so it seems like keeping this cast in there should be reasonable and keep things safe.

Comment (kratsg) [reply src/pyhf/tensor/tensorflow_backend.py]: > as a valid form of Bool (as NumPy does)

It's not that numpy handles it as a valid form, it type-casts for you implicitly as part of the masking - but that might not be good behavior. We should have users be explicit about the type of tensor they're dealing with (bools, floats, ints) - because they all store memory in different ways and are allocated differently.

> The cast should be faster then the `self.astensor` call that we have previously been doing

Then we should change the tensorflow `astensor` to use the `cast` then... if it's faster.

File: src/pyhf/tensor/common.py (hunk 2 above)
Comment (matthewfeickert) [reply src/pyhf/tensor/common.py]: Should be resolved now.

File: src/pyhf/tensor/tensorflow_backend.py (hunk 7 above)
Comment (matthewfeickert) [reply src/pyhf/tensor/tensorflow_backend.py]: > Then we should change the tensorflow `astensor` to use the `cast` then... if it's faster.

We do already

https://github.com/diana-hep/pyhf/blob/f952b8bca853e09259351b2ebc2a09c918c23e6f/src/pyhf/tensor/tensorflow_backend.py#L152-L153

doing this just skips the rest of the logic of `astensor`. I will still work on changing the calls in other parts, as needed, to explicitly request Bools.

File: src/pyhf/interpolators/code0.py [hunk 8]
```diff
@@ -71,8 +71,11 @@ def __call__(self, alphasets):
             'sa,shb->shab', alphasets, self.deltas_dn
         )
 
-        masks = tensorlib.einsum(
-            'sa,shb->shab', where_alphasets_positive, self.broadcast_helper
+        masks = tensorlib.astensor(
```
Comment (lukasheinrich) [inline src/pyhf/interpolators/code0.py]: doesn't einsum return a tensor anyways?

File: src/pyhf/modifiers/lumi.py [hunk 9]
```diff
@@ -69,4 +69,8 @@ def apply(self, pars):
         else:
             results_lumi = tensorlib.einsum('msab,xa->msab', self.lumi_mask, lumis)
 
-        return tensorlib.where(self.lumi_mask, results_lumi, self.lumi_default)
+        return tensorlib.where(
+            tensorlib.astensor(self.lumi_mask, dtype="bool"),
```
Comment (lukasheinrich) [inline src/pyhf/modifiers/lumi.py]: this can probably go in the _precompute of the modifier where `self.lumi_mask` is set up

File: src/pyhf/modifiers/normfactor.py [hunk 10]
```diff
@@ -74,6 +73,8 @@ def apply(self, pars):
             )
 
         results_normfactor = tensorlib.where(
-            self.normfactor_mask, results_normfactor, self.normfactor_default
+            tensorlib.astensor(self.normfactor_mask, dtype="bool"),
```
Comment (lukasheinrich) [inline src/pyhf/modifiers/normfactor.py]: same here.. could just wrap the `tensorlib.tile` call in precompute to do the cast

File: src/pyhf/modifiers/shapefactor.py [hunk 11]
```diff
@@ -118,7 +118,7 @@ def _precompute(self):
         if not self.param_viewer.index_selection:
             return
         tensorlib, _ = get_backend()
-        self.shapefactor_mask = tensorlib.astensor(self._shapefactor_mask)
+        self.shapefactor_mask = tensorlib.astensor(self._shapefactor_mask, dtype="bool")
```
Comment (lukasheinrich) [inline src/pyhf/modifiers/shapefactor.py]: same here.. could just wrap the tensorlib.tile call in precompute to do the cast

File: src/pyhf/interpolators/code0.py (hunk 8 above)
Comment (matthewfeickert) [reply src/pyhf/interpolators/code0.py]: @lukasheinrich It does. Dumb question: Do we support a way to cast the return Tensor from `tensorlib`?

File: src/pyhf/modifiers/lumi.py (hunk 9 above)
Comment (matthewfeickert) [reply src/pyhf/modifiers/lumi.py]: @lukasheinrich I looked at this again and I don't think this will work. If `einsum` is given `self.lumi_mask` as a Bool tensor then it will crash as the backends don't support `Bool` there. Sorry, I ran into this last night but didn't mention it anywhere.

Comment (lukasheinrich) [reply src/pyhf/modifiers/lumi.py]: ok I guess we could have `self.lumi_mask_bool`, and `self.lumi_mask` and use each at their appropriate position, but probably doesn't make that much of a diff

File: src/pyhf/modifiers/shapefactor.py (hunk 11 above)
Comment (matthewfeickert) [reply src/pyhf/modifiers/shapefactor.py]: @lukasheinrich I simplified this into one call instead of two, but this still needs to be done as `tile` requires that `self.shapefactor_mask` is a Tensor, so this needs to be done before `tile` gets evaluated.

File: src/pyhf/modifiers/lumi.py (hunk 9 above)
Comment (matthewfeickert) [reply src/pyhf/modifiers/lumi.py]: I can do this.
